File: trunk/java/core/src/org/daisy/urakawa/media/AbstractMedia.java
package org.daisy.urakawa.media;

import java.net.URI;

import org.daisy.urakawa.AbstractXukAbleWithPresentation;
import org.daisy.urakawa.FactoryCannotCreateTypeException;
import org.daisy.urakawa.Presentation;
import org.daisy.urakawa.events.DataModelChangedEvent;
import org.daisy.urakawa.events.Event;
import org.daisy.urakawa.events.EventHandler;
import org.daisy.urakawa.events.IEventHandler;
import org.daisy.urakawa.events.IEventListener;
import org.daisy.urakawa.events.LanguageChangedEvent;
import org.daisy.urakawa.exception.IsNotInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.nativeapi.IXmlDataReader;
import org.daisy.urakawa.nativeapi.IXmlDataWriter;
import org.daisy.urakawa.progress.IProgressHandler;
import org.daisy.urakawa.progress.ProgressCancelledException;
import org.daisy.urakawa.xuk.XukDeserializationFailedException;
import org.daisy.urakawa.xuk.XukSerializationFailedException;

/**
 * 
 *
 */
public abstract class AbstractMedia extends AbstractXukAbleWithPresentation
        implements IMedia
{
    protected IEventListener<DataModelChangedEvent> mBubbleEventListener = new IEventListener<DataModelChangedEvent>()
    {
        public <K extends DataModelChangedEvent> void eventCallback(K event)
                throws MethodParameterIsNullException
        {
            if (event == null)
            {
                throw new MethodParameterIsNullException();
            }
            notifyListeners(event);
        }
    };
    protected IEventHandler<Event> mLanguageChangedEventNotifier = new EventHandler();
    protected IEventHandler<Event> mDataModelEventNotifier = new EventHandler();

    public <K extends DataModelChangedEvent> void notifyListeners(K event)
            throws MethodParameterIsNullException
    {
        if (LanguageChangedEvent.class.isAssignableFrom(event.getClass()))
        {
            mLanguageChangedEventNotifier.notifyListeners(event);
        }
        mDataModelEventNotifier.notifyListeners(event);
    }

    public <K extends DataModelChangedEvent> void registerListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (LanguageChangedEvent.class.isAssignableFrom(klass))
        {
            mLanguageChangedEventNotifier.registerListener(listener, klass);
        }
        else
        {
            mDataModelEventNotifier.registerListener(listener, klass);
        }
    }

    public <K extends DataModelChangedEvent> void unregisterListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (LanguageChangedEvent.class.isAssignableFrom(klass))
        {
            mLanguageChangedEventNotifier.unregisterListener(listener, klass);
        }
        else
        {
            mDataModelEventNotifier.unregisterListener(listener, klass);
        }
    }

    /**
	 * 
	 */
    public AbstractMedia()
    {
        super();
        mLanguage = null;
    }

    private String mLanguage;

    public abstract boolean isContinuous();

    public abstract boolean isDiscrete();

    public abstract boolean isSequence();

    public AbstractMedia copy()
    {
        return (AbstractMedia) copyProtected();
    }

    protected IMedia copyProtected()
    {
        IMedia expMedia;
        try
        {
            expMedia = getPresentation().getMediaFactory().create(
                    getXukLocalName(), getXukNamespaceURI());
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        catch (IsNotInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        try
        {
            expMedia.setLanguage(getLanguage());
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return expMedia;
    }

    public IMedia export(Presentation destPres)
            throws FactoryCannotCreateTypeException,
            MethodParameterIsNullException
    {
        if (destPres == null)
        {
            throw new MethodParameterIsNullException();
        }
        return exportProtected(destPres);
    }

    protected IMedia exportProtected(Presentation destPres)
            throws FactoryCannotCreateTypeException,
            MethodParameterIsNullException
    {
        if (destPres == null)
        {
            throw new MethodParameterIsNullException();
        }
        IMedia expMedia;
        try
        {
            expMedia = destPres.getMediaFactory().create(getXukLocalName(),
                    getXukNamespaceURI());
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        if (expMedia == null)
        {
            throw new FactoryCannotCreateTypeException();
        }
        try
        {
            expMedia.setLanguage(getLanguage());
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return expMedia;
    }

    public void setLanguage(String lang)
            throws MethodParameterIsEmptyStringException
    {
        if (lang.length() == 0)
        {
            throw new MethodParameterIsEmptyStringException();
        }
        String prevlang = mLanguage;
        mLanguage = lang;
        if (prevlang != mLanguage)
            try
            {
                notifyListeners(new LanguageChangedEvent(this, mLanguage,
                        prevlang));
            }
            catch (MethodParameterIsNullException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
    }

    public String getLanguage()
    {
        return mLanguage;
    }

    @Override
    protected void clear()
    {
        mLanguage = null;
        // super.clear();
    }

    @Override
    protected void xukInAttributes(IXmlDataReader source, IProgressHandler ph)
            throws MethodParameterIsNullException,
            XukDeserializationFailedException, ProgressCancelledException
    {
        if (source == null)
        {
            throw new MethodParameterIsNullException();
        }
        // To avoid event notification overhead, we bypass this:
        if (false && ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        String lang = source.getAttribute("language");
        if (lang != null)
        {
            lang = lang.trim();
            if (lang.length() == 0)
            {
                lang = null;
            }
        }
        try
        {
            setLanguage(lang);
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        // super.xukInAttributes(source, ph);
    }

    @Override
    protected void xukOutAttributes(IXmlDataWriter destination, URI baseUri,
            IProgressHandler ph) throws MethodParameterIsNullException,
            XukSerializationFailedException, ProgressCancelledException
    {
        if (ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        if (getLanguage() != null)
            destination.writeAttributeString("language", getLanguage());
        // super.xukOutAttributes(destination, baseUri, ph);
    }

    public boolean ValueEquals(IMedia other)
            throws MethodParameterIsNullException
    {
        if (other == null)
            throw new MethodParameterIsNullException();
        if (this.getClass() != other.getClass())
            return false;
        if (this.getLanguage() != other.getLanguage())
            return false;
        return true;
    }
}

File: tags/1.1_Obi-1.0/csharp/core/media/AbstractMedia.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace urakawa.media
{
	/// <summary>
	/// Abstract implementation of <see cref="IMedia"/> - used as bvase of all <see cref="IMedia"/>
	/// </summary>
	public abstract class AbstractMedia : WithPresentation, IMedia
	{
		#region Event related members

		/// <summary>
		/// Event fired after the <see cref="TextMedia"/> has changed. 
		/// The event fire before any change specific event 
		/// </summary>
		public event EventHandler<urakawa.events.DataModelChangedEventArgs> changed;
		/// <summary>
		/// Fires the <see cref="changed"/> event 
		/// </summary>
		/// <param name="args">The arguments of the event</param>
		protected void notifyChanged(urakawa.events.DataModelChangedEventArgs args)
		{
			EventHandler<urakawa.events.DataModelChangedEventArgs> d = changed;
			if (d != null) d(this, args);
		}

		/// <summary>
		/// Event fired after the language of the <see cref="TextMedia"/> has changed
		/// </summary>
		public event EventHandler<urakawa.events.LanguageChangedEventArgs> languageChanged;
		/// <summary>
		/// Fires the <see cref="languageChanged"/> event
		/// </summary>
		/// <param name="source">The source, that is the <see cref="TextMedia"/> whoose language changed</param>
		/// <param name="newLang">The new value for the language</param>
		/// <param name="prevLang">The value for the language prior to the change</param>
		protected void notifyLanguageChanged(AbstractMedia source, string newLang, string prevLang)
		{
			EventHandler<urakawa.events.LanguageChangedEventArgs> d = languageChanged;
			if (d != null) d(this, new urakawa.events.LanguageChangedEventArgs(source, newLang, prevLang));
		}

		void this_languageChanged(object sender, urakawa.events.LanguageChangedEventArgs e)
		{
			notifyChanged(e);
		}
		#endregion

		/// <summary>
		/// Default constructor
		/// </summary>
		public AbstractMedia()
		{
			mLanguage = null;
			this.languageChanged += new EventHandler<urakawa.events.LanguageChangedEventArgs>(this_languageChanged);
		}

		private string mLanguage;

		#region IMedia Members


		/// <summary>
		/// Gets the <see cref="IMediaFactory"/> associated with the <see cref="AbstractMedia"/> (via. the owning <see cref="Presentation"/>
		/// </summary>
		/// <returns>The <see cref="IMediaFactory"/></returns>
		public IMediaFactory getMediaFactory()
		{
			return getPresentation().getMediaFactory();
		}

		/// <summary>
		/// Determines if the <see cref="AbstractMedia"/> is continuous
		/// </summary>
		/// <returns>A <see cref="bool"/> indicating if the <see cref="AbstractMedia"/> is continuous</returns>
		public abstract bool isContinuous();

		/// <summary>
		/// Determines if the <see cref="AbstractMedia"/> is discrete
		/// </summary>
		/// <returns>A <see cref="bool"/> indicating if the <see cref="AbstractMedia"/> is discrete</returns>
		public abstract bool isDiscrete();

		/// <summary>
		/// Determines if the <see cref="AbstractMedia"/> is a <see cref="SequenceMedia"/>
		/// </summary>
		/// <returns>A <see cref="bool"/> indicating if the <see cref="AbstractMedia"/> is a <see cref="SequenceMedia"/></returns>
		public abstract bool isSequence();

		IMedia IMedia.copy()
		{
			return copyProtected();
		}

		/// <summary>
		/// Creates a copy of the <see cref="AbstractMedia"/>
		/// </summary>
		/// <returns>The copy</returns>
		public AbstractMedia copy()
		{
			return copyProtected() as AbstractMedia;
		}

		/// <summary>
		/// Creates a copy of the <see cref="AbstractMedia"/>
		/// </summary>
		/// <returns>The copy</returns>
		protected virtual IMedia copyProtected()
        {
            IMedia expMedia = getPresentation().getMediaFactory().createMedia(getXukLocalName(), getXukNamespaceUri());
            if (expMedia == null)
            {
                throw new exception.FactoryCannotCreateTypeException(String.Format(
                    "The factory of the destination Presentation cannot create a AbstractMedia matching Xuk QName {1}:{0}",
                    getXukLocalName(), getXukNamespaceUri()));
            }
            expMedia.setLanguage(getLanguage());
            return expMedia;
		}

		IMedia IMedia.export(Presentation destPres)
		{
			return exportProtected(destPres);
		}

		/// <summary>
		/// Exports the <see cref="AbstractMedia"/> to a given destination <see cref="Presentation"/>
		/// </summary>
		/// <param name="destPres">The destination <see cref="Presentation"/></param>
		/// <returns>The exported <see cref="AbstractMedia"/></returns>
		public AbstractMedia export(Presentation destPres)
		{
			return exportProtected(destPres) as AbstractMedia;
		}

		/// <summary>
		/// Exports the <see cref="AbstractMedia"/> to a given destination <see cref="Presentation"/>
		/// </summary>
		/// <param name="destPres">The destination <see cref="Presentation"/></param>
		/// <returns>The exported <see cref="AbstractMedia"/></returns>
		protected virtual IMedia exportProtected(Presentation destPres)
		{
			AbstractMedia expMedia = destPres.getMediaFactory().createMedia(getXukLocalName(), getXukNamespaceUri()) as AbstractMedia;
			if (expMedia == null)
			{
				throw new exception.FactoryCannotCreateTypeException(String.Format(
					"The factory of the destination Presentation cannot create a AbstractMedia matching Xuk QName {1}:{0}",
					getXukLocalName(), getXukNamespaceUri()));
			}
			expMedia.setLanguage(getLanguage());
			return expMedia;
		}

		/// <summary>
		/// Sets the language of the <see cref="AbstractMedia"/>
		/// </summary>
		/// <param name="lang">The new language, can be null but not an empty string</param>
		/// <exception cref="exception.MethodParameterIsEmptyStringException">
		/// Thrown if the new language is an empty <see cref="String"/></exception>
		public void setLanguage(string lang)
		{
			if (lang == "")
			{
				throw new exception.MethodParameterIsEmptyStringException(
					"The language can not be an empty string");
			}
			string prevlang = mLanguage;
			mLanguage = lang;
			if (prevlang != mLanguage) notifyLanguageChanged(this, mLanguage, prevlang);
		}

		/// <summary>
		/// Gets the language of the <see cref="AbstractMedia"/>
		/// </summary>
		/// <returns>The language</returns>
		public string getLanguage()
		{
			return mLanguage;
		}

		#endregion

		/// <summary>
		/// Clears the data of the <see cref="AbstractMedia"/>
		/// </summary>
		protected override void clear()
		{
			mLanguage = null;
			base.clear();
		}


		/// <summary>
		/// Reads the attributes of a AbstractMedia xuk element.
		/// </summary>
		/// <param name="source">The source <see cref="XmlReader"/></param>
		protected override void xukInAttributes(XmlReader source)
		{
			string lang = source.GetAttribute("language");
			if (lang != null) lang = lang.Trim();
			if (lang == "") lang = null;
			setLanguage(lang);
			base.xukInAttributes(source);
		}

		/// <summary>
		/// Writes the attributes of a AbstractMedia element
		/// </summary>
		/// <param name="destination">The destination <see cref="XmlWriter"/></param>
		/// <param name="baseUri">
		/// The base <see cref="Uri"/> used to make written <see cref="Uri"/>s relative, 
		/// if <c>null</c> absolute <see cref="Uri"/>s are written
		/// </param>
		protected override void xukOutAttributes(XmlWriter destination, Uri baseUri)
		{
			if (getLanguage()!=null) destination.WriteAttributeString("language", getLanguage());

			base.xukOutAttributes(destination, baseUri);
		}

		#region IValueEquatable<IMedia> Members

		/// <summary>
		/// Compares <c>this</c> with a given other <see cref="IMedia"/> for equality
		/// </summary>
		/// <param name="other">The other <see cref="IMedia"/></param>
		/// <returns><c>true</c> if equal, otherwise <c>false</c></returns>
		public virtual bool valueEquals(IMedia other)
		{
			AbstractMedia amOther = other as AbstractMedia;
			if (amOther == null) return false;
			if (this.GetType() != amOther.GetType()) return false;
			if (this.getLanguage() != amOther.getLanguage()) return false;
			return true;
		}

		#endregion
	}
}
